### ISSUE TITLE
Get bkt into the major package managers

### ISSUE THREAD

**Issue by dimo414** [dimo414/bkt#12]
I don't have much experience with this, but filing a bug to keep track of this over time. Package managers to consider: https://github.com/junegunn/fzf#using-linux-package-managers.

* [ ] Arch / Pacman
* [ ] Debian / `apt-get`
* [ ] macOS / Homebrew

**Comment by thiagowfx** [comment]
Now packaged for Arch: https://aur.archlinux.org/packages/bkt/

**Comment by dimo414** [comment]
Thanks a lot @thiagowfx! And props for your [nice writeup](https://thiagowfx.github.io/2022/01/arch-linux-new-pkgbuild-workflow/) as well.

**Comment by dimo414** [comment]
Thank you very much! It looks like it's already showing up in the README :)

**Comment by 0323pin** [comment]
Thank you for this tool. I'm using it in a slightly unconventional way, though 😄 
I'm setting the time to few minutes and launching troublesome programs to catch why and when they crash or segfault.

Yeah, if you pulling the info from repology, it will mostly be up-to-date.
I say mostly because we use `cvs` and the git-repository is just a mirror. Repology pulls information from git and it can be a delay of up to one day. So, don't worry if you make a new release and it takes a day or two to be marked as updated, that's normal.

**Comment by dimo414** [comment]
> setting the time to few minutes and launching troublesome programs to catch why and when they crash or segfault.

Hey that sounds like a pretty reasonable use case to me :) If you'd like to share how you're using `bkt` in more detail in https://github.com/dimo414/bkt/discussions it might inspire other or help me plan future features.

**Comment by dimo414** [comment]
Hah alright, that is a bit more unusual :) Happy to chat about it more in a discussion if you'd like to post one.

**Comment by 0323pin** [comment]
@dimo414 Forgot to reply. Do you think what I'm doing is wrong? Of course `/tmp` is cleared upon reboot but, I could re-direct it to a _more permanent_ location.

**Comment by ehaupt** [comment]
I've just packaged `bktr` for FreeBSD: https://cgit.freebsd.org/ports/commit/?id=6f16c9c835c7c33e302e5e2b2a559a00d210944d

**Comment by benthetechguy** [comment]
Wow, that's some cool timing! I just began work packaging `bkt` for Debian almost exactly when you made that comment.

**Comment by dimo414** [comment]
Amazing, thank you @ehaupt! And thank you in advance @benthetechguy :)

**Comment by ehaupt** [comment]
Maybe we can convince @alexmyczko to create a Homebrew receipe :stuck_out_tongue_winking_eye:

**Comment by zbentley** [comment]
Homebrew recipe is here: https://github.com/Homebrew/homebrew-core/pull/95407

However, that won't be merged by their maintainers until it passes their popularity test (error message: `GitHub repository not notable enough (<30 forks, <30 watchers and <75 stars)`), so it's in draft for now.

**Comment by zbentley** [comment]
The Homebrew PR merged:

```
zac.bentley@ZacBentleyMBP ~ ∴ brew update; brew install bkt
==> Pouring bkt--0.5.2.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/bkt/0.5.2: 7 files, 1.3MB
zac.bentley@ZacBentleyMBP ~ ∴ bkt
error: The following required arguments were not provided:
    <COMMAND>...

USAGE:
    bkt [OPTIONS] [--] <COMMAND>...

For more information try --help
```

**Comment by dimo414** [comment]
Thanks a lot @zbentley!

**Comment by dimo414** [comment]
@benthetechguy just curious if you had any luck getting `bkt` packaged for Debian?

**Comment by benthetechguy** [comment]
The way Debian does rust packaging is kind of confusing. The source code of all the dependencies and their dependencies need to be packaged so they can be compiled at build time of the program.
I had the actual packaging of bkt done an hour after I made my initial comment, but there's still work going on to package all the dependencies and get them uploaded into the Debian archive, which can take a few weeks.

**Comment by dimo414** [comment]
Dang, I did not know that. What dependencies are you having trouble with? I'm not opposed to dropping/swapping dependencies that are a pain. Thank you for digging into this!

**Comment by benthetechguy** [comment]
It would help if the clap dependency was 2 instead of 3; the person who's working on clap is having a hard time when it comes to upgrading our version of clap to 3.

**Comment by 0323pin** [comment]
> The way Debian does rust packaging is kind of confusing. The source code of all the dependencies and their dependencies need to be packaged so they can be compiled at build time of the program.

So they did this in the end, I remember reading the discussion about it but, never checked back on the decision. Is Debian doing the same with Go modules?

I just hope this idea doesn't spread, it might be the most correct way of packaging but, it's unsustainable, it uses a huge amount of resources and must be a nightmare for package maintainers. _E.g._ I keep ~ 70 Rust packages up-to-date on pkgsrc, some with a really long list of crate dependencies, IRC the largest one has 520 crate dependecies, that would be 36 400 packages to keep track of, in case all of them were the same size, which is not the case of course.

In any case, keeping my fingers crossed that this remains a Debian only thing 😨

**Comment by dimo414** [comment]
Hah that's a bummer, @tranzystorek-io swapped us to clap 3 in #1. We can definitely remove the patch version from serde-json.

**Comment by tranzystorek-io** [comment]
I've started the process to add `bkt` to Void Linux: https://github.com/void-linux/void-packages/pull/36124

**Comment by benthetechguy** [comment]
> Hah that's a bummer, @tranzystorek-io swapped us to clap 3 in https://github.com/dimo414/bkt/pull/1. We can definitely remove the patch version from serde-json.

That's good about serde-json. Can anything be done about clap? Are those features relying on clap 3 an integral part of the codebase? If so, it'll probably take much longer as we'll have to wait for the Debian clap maintainer to finish clap 3.

**Comment by dimo414** [comment]
I'm hesitant to revert back to clap 2 simply because we already went through the trouble to upgrade, but I'm curious what @tranzystorek-io thinks. My impression is there's sufficient wind behind clap-3's sails that a Debian release can't be _that_ far behind, but I'm not well informed on the matter. Is there a tracking issue or more context you can point to wrt clap-3 on Debian?

**Comment by benthetechguy** [comment]
#debian-rust on OFTC is where all the discussion happens, and the work being done is on [this repo](https://salsa.debian.org/rust-team/debcargo-conf.git).
There is one interesting thing happening, however. A package `rust-clap-2` was just uploaded to Debian a few minutes ago. When a package is upgraded but others still depend on the old version, this can happen. The uploading of `rust-clap-2` is a hint that `rust-clap` will be upgraded to 3 in the near future. This gives me hope.
I've just asked what the current status of clap 3 is.

**Comment by tranzystorek-io** [comment]
Reverting to pure clap 2 would indeed be counterproductive and non-trivial.

That said, we could switch back fairly easily to structopt which relies on clap 2.

@benthetechguy I'm guessing structopt is available on the debian side already?

**Comment by benthetechguy** [comment]
Yeah, it is. I'd wait a bit before changing anything too much, though; It's possible that if you wait a little bit we'll have clap 3 in Debian. As mentioned previously, we've had a hint that clap 3 is around the corner, and I'm awaiting a response from the maintainer for a status report.

**Comment by tranzystorek-io** [comment]
Nice, the serde version drop also bumped up the autocfg version, which fixes an [issue](https://github.com/void-linux/void-packages/issues/34889) in Void CI when building cross targets 🎉

**Comment by benthetechguy** [comment]
Good news, clap 3 is now in Debian. This means I should be able to package bkt, and provided someone's there to sponsor the upload, get it into Debian!

**Comment by alexmyczko** [comment]
@benthetechguy I will be glad to review and sponsor bkt.

**Comment by benthetechguy** [comment]
It seems that the derive feature is missing from Debian's clap. I will investigate the cause for this. In the meantime, in case getting derive is a no-go, how important is it to bkt?

**Comment by tranzystorek-io** [comment]
I rewrote all clap v2 parsing into clap deriv to make the code more readable and concise, going back to the builder API is a step backwards. 😐

**Comment by tranzystorek-io** [comment]
@dimo414 I'm not sure if this is worth the time but in the Void CI the tests sometimes fail, somewhat nondeterministically: https://github.com/void-linux/void-packages/runs/5711656703?check_suite_focus=true

**Comment by dimo414** [comment]
Thanks for the heads up, let's follow up in #14.

**Comment by benthetechguy** [comment]
Good news and bad news.
The good news is that I have solved the clap_derive issue and it will be in Debian soon.
The bad news is that there are problems with test_dir. The code is quite outdated and has some problems compiling in Debian's setup. Can we do without test_dir, or is it absolutely needed for bkt?

**Comment by dimo414** [comment]
I can definitely swap off test_dir to something else if needed, it's just used to simplify the test setup. I asked the [crate maintainer if they can release an update](https://github.com/rpacholek/test_dir/issues/1) and if not we can drop that dependency.

**Comment by benthetechguy** [comment]
Perfect, thanks. I was worried because it seemed to be everywhere in the
code. I'm not experienced with Rust itself, so I'm not good at seeing how
important something is outside of the size of the code itself.
When I get access to my machine again, I'll post the exact error(s) I'm
getting in the issue you linked above.

**Comment by dimo414** [comment]
Yeah it's used in pretty much every test, but it's not compiled into the production binary. You can see that because it's under [dev-dependencies](https://github.com/dimo414/bkt/blob/d95dfe10401d1bbbea774356e470687d6ac94606/Cargo.toml#L37-L39) (if you know Rust of course 😄).

**Comment by benthetechguy** [comment]
Yes, Debian requires dev-dependencies for autopkgtests.

**Comment by dimo414** [comment]
Right sorry, I was just saying it doesn't impact the behavior of the binary itself, meaning to imply that makes it relatively easy to swap out if needed.

**Comment by benthetechguy** [comment]
I think we might need to end up doing that. The maintainer of test_dir doesn't seem to be very active.

**Comment by dimo414** [comment]
Published [0.5.4](https://github.com/dimo414/bkt/releases/tag/0.5.4)

**Comment by benthetechguy** [comment]
Thanks. bkt should now be ready for Debian packaging as soon as someone uploads clap_derive and clap.
The reason things are often outdated in Debian is because so much time is spent waiting around for someone to bother to upload something ;)

P.S: You don't need to make a new release every time you change dependencies; we can and will patch trivial things.

**Comment by benthetechguy** [comment]
I have created a package if anyone wants to test: [bkt_0.5.4-1_arm64.deb.gz](https://github.com/dimo414/bkt/files/8421687/bkt_0.5.4-1_arm64.deb.gz)
It's compressed with gzip because for some reason GitHub only allows certain file types to be uploaded and this is one of them.

**Comment by alexmyczko** [comment]
@benthetechguy I'd like to review a debian source package, if someone would like to have it officially in Debian (well I would like so)

**Comment by benthetechguy** [comment]
It's a bit complicated for Debian Rust packaging. Information is available at https://salsa.debian.org/rust-team/debcargo-conf

**Comment by benthetechguy** [comment]
Dependencies were uploaded by @sylvestre about a week ago, and all but one just made it through the queue a couple hours ago. Since the dependencies are pretty much done, this means that bkt itself should be ready for upload as soon as everything clears.

**Comment by benthetechguy** [comment]
Last dependency just cleared 🥳

**Comment by benthetechguy** [comment]
After a month, a Debian Member finally bothered to upload bkt, and it has now been accepted into Debian unstable. If there are no bugs, it should migrate to testing after a couple days.
https://packages.debian.org/sid/bkt

**Comment by dimo414** [comment]
Thanks for the update! That link seems broken now, did it get rolled back?

**Comment by benthetechguy** [comment]
It seems that the autopkgtests have failed due to a dependency that hasn't yet gone through the upload queue. This should resolve itself after a few days.
You can already install bkt on amd64 on Debian unstable: https://deb.debian.org/debian/pool/main/r/rust-bkt/bkt_0.5.4-1_amd64.deb

**Comment by dimo414** [comment]
It seems like bkt is still stuck in unstable, perhaps because of the flaky test tracked in #14? I've just disabled that test for the time being, in case that's helpful.

**Comment by tranzystorek-io** [comment]
Aaaand we're on Void now! (Took the liberty of disabling that one test in CI)

**Comment by dimo414** [comment]
Neat, thank you!

**Comment by thiagowfx** [comment]
NixOS packaging would also be nice!

**Comment by dimo414** [comment]
@benthetechguy do you happen to know what's needed to get bkt out of Debian unstable?

Also FYI to everyone who is so kindly maintaining packages, I plan to cut a 0.6.0 release in the next few days.

**Comment by alexmyczko** [comment]
Not sure but fixing this? https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1026801

And no worries, people can request a backport of it...

**Comment by dimo414** [comment]
@alexmyczko sorry I missed your reply; I _think_ that failure has been resolved by 61da09ae56. If it hasn't it would be great to report an issue here, ideally with reproduction steps, since I don't see bug reports submitted against Debian.

**Comment by benthetechguy** [comment]
Hi,

Sorry for my long inactivity. I've been busy with school and haven't had 
time to work much on bkt packaging or Debian as a whole. I haven't used 
bkt for a while now and thus no longer have much of a motivation to work 
on its packaging, so I would much appreciate if someone else would adopt 
the package.

Since it's already part of the Rust Team, you don't have to do anything 
special apart from adding yourself to the Uploaders list in 
debcargo.toml (feel free to also remove me). I think most of the hard 
work is done and all you have to do to get this package to migrate 
properly to testing and stable from there is to update to the latest 
version. Unfortunately, since we are only a few weeks away from the 
release of Debian 12 Bookworm, it doesn't seem like the new version 
would make it in time to reach the stable release of Debian.

Thanks and sorry again!

**Comment by alexmyczko** [comment]
i have requested BPO of bkt in #debian-rust

**Comment by alexanderkjall** [comment]
bkt is now packaged for trixie: https://packages.debian.org/trixie/bkt

**Comment by dimo414** [comment]
That's great to hear, thank you so much!

**Comment by yonas** [comment]
Alpine request here: https://gitlab.alpinelinux.org/alpine/aports/-/issues/15731

**Comment by thiagowfx** [comment]
Alpine: https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/68650
